lib/src/matrix.dart

```diff
@@ -1,86 +1,83 @@
 import 'package:ml_linalg/vector.dart';
 
+import '../df.dart';
+
 /// A class to manage the data inside the [DataFrame]
 class DataMatrix {
   /// The dataset
-  List<List<dynamic>> data = <List<dynamic>>[];
+  List<List<Object?>> data = [];
 
   // ********* insert operations **********
 
   /// Add a row
-  void addRow(Map<String, dynamic> row, Map<int, String> indices) {
-    //print("DF ADD ROW $row / $indices");
-    final r = <dynamic>[];
-    var i = 0;
-    row.forEach((k, dynamic v) {
-      final keyName = indices[i];
-      r.add(row[keyName]);
-      ++i;
-    });
-    data.add(r);
-  }
+  void addRow(Map<String, Object?> row, List<String> columnNames) =>
+      data.add(columnNames.map((colName) => row[colName]).toList());
 
   // ********* select operations **********
 
   /// Row for an index position
-  Map<String, dynamic> rowForIndex(int index, Map<int, String> indices) {
-    final row = <String, dynamic>{};
+  Map<String, Object> rowForIndex(
+      int index, List<String> indicesToColumnNames) {
+    final row = <String, Object>{};
     final dataRow = data[index];
     var i = 0;
-    dataRow.forEach((dynamic item) {
-      row[indices[i]] = item;
-      ++i;
+    dataRow.forEach((item) {
+      if (item != null) row[indicesToColumnNames[i]] = item;
+      i++;
     });
     return row;
   }
 
   /// Rows for an index range of positions
-  List<Map<String, dynamic>> rowsForIndexRange(
-      int startIndex, int endIndex, Map<int, String> indices) {
-    final dataRows = <Map<String, dynamic>>[];
+  List<Map<String, Object>> rowsForIndexRange(
+      int startIndex, int endIndex, List<String> columnNames) {
+    final dataRows = <Map<String, Object>>[];
     for (final row in data.sublist(startIndex, endIndex)) {
-      final dataRow = <String, dynamic>{};
+      final dataRow = <String, Object>{};
       var i = 0;
-      row.forEach((dynamic item) {
-        dataRow[indices[i]] = item;
-        ++i;
+      row.forEach((Object? item) {
+        if (item != null) dataRow[columnNames[i]] = item;
```

Comment:
Rather than returning a map with nulls, we simply return a map with null entries unspecified. This means we have no way to distinguish between a value that was never specified or a value that was explicitly specified as null. If we want to change this this will require us adding an extra sentinel type for "not specified" or something-this would be a significant refactor. Let me know if this is desirable.